File: Niveau 3/Niveau 3.11/labyrinthe.py
#-------------------------------------------------------------------------------
# Name:        Labyrinthe à billes
# Purpose:     France-IOI niveau 3.11
#
# Author:      Sébastien
#
# Created:     31/10/2022
#-------------------------------------------------------------------------------

from operator import itemgetter

mur   = '#'
vide  = '.'
trou  = 'O'
bille = 'x'
deplacements = {'N': (-1,0), 'S': (+1,0), 'O': (0,-1), 'E': (0,+1)}

def deplacer(labyrinthe, position, direction):
    deplacement = deplacements.get(direction)
    if deplacement == None: return position
    labyrinthe[position[0]][position[1]] = '.'

    while True:
        j, i = position[0] + deplacement[0], position[1] + deplacement[1]
        if labyrinthe[j][i] == trou:
            return None
        elif labyrinthe[j][i] == mur or labyrinthe[j][i] == bille:
            labyrinthe[position[0]][position[1]] = bille
            return position
        else:
            position = (j, i)

def main():
    l, c = map(int, input().split())
    labyrinthe = [[c for c in input()] for _ in range(l)]
    billes = []
    for j in range(l):
        for i in range(c):
            if labyrinthe[j][i] == bille:
                billes.append((j, i))

    _ = int(input())
    deplacements = input()
    for d in deplacements:
        if   d == 'N': billes.sort(key = itemgetter(0))
        elif d == 'S': billes.sort(key = itemgetter(0), reverse=True)
        elif d == 'O': billes.sort(key = itemgetter(1))
        elif d == 'E': billes.sort(key = itemgetter(1), reverse=True)
##        print("Avant déplacement :", billes)
        billes = [b for b in map(lambda x: deplacer(labyrinthe, x, d), billes) if b != None]
##        print("Après déplacement :", billes)

    print(len(billes))

if __name__ == '__main__':
    main()

"""
7 8
########
#x.....#
#....O.#
#....#.#
####.#.#
#x...#.#
########
4
SENE
---
7 8
########
#x...x.#
#....O.#
#....#.#
#O##.#.#
#x...#.#
########
4
SE NE
"""

File: Niveau 3/Niveau 3.11/labyrinthe.hs
--------------------------------------------------------------------------------
-- Name:        Labyrinthe à billes
-- Purpose:     France-IOI niveau 3.11
--
-- Author:      Sébastien
--
-- Created:     01/11/2022
--------------------------------------------------------------------------------

import Data.Array.Unboxed
import Data.List
import Control.Monad
import qualified Data.Map as M

type Coord = (Int, Int)
type Plan = Array Coord Char

mur   = '#'
vide  = '.'
trou  = 'O'
bille = 'x'
deplacements = M.fromList[('N', (-1,0)), ('S', (1,0)), ('O', (0,-1)), ('E', (0,1))   ]

main :: IO ()
main = do
    (l:c:_) <- fmap (map read . words) getLine :: IO [Int]
    contenu <- fmap (filter (flip elem [mur, vide, trou, bille]) . unlines) $ replicateM l getLine
    let labyrinthe = listArray ((1,1), (l,c)) contenu :: Plan
        is = indices labyrinthe
        billes = foldr (\i bs -> if labyrinthe!i == bille then i:bs else bs) [] is
        labyrinthe' = labyrinthe // zip billes (repeat vide)
    putStrLn $ show billes
    _ <- readLn :: IO Int
    directions <- getLine
    billes' <- deplacerBilles labyrinthe directions billes
    print $ length billes'

-- Enchaîne une série de déplacement
deplacerBilles :: Plan -> String -> [Coord] -> IO [Coord]
deplacerBilles _ [] billes = return billes
deplacerBilles labyrinthe (direction:ds) billes = do
    let billes' = deplacerBilles' labyrinthe direction billes
    putStrLn $ show billes ++ " -> " ++ direction:"" ++ " -> " ++ show billes'
    deplacerBilles labyrinthe ds billes'

-- Déplace toutes les billes dans une direction
deplacerBilles' :: Plan -> Char -> [Coord] -> [Coord]
deplacerBilles' labyrinthe direction billes =
    case M.lookup direction deplacements of
        -- Just deplacement -> map (deplacerBille labyrinthe deplacement billes) billes
        Just deplacement -> let billes' = trierBilles direction billes
                            in foldr (\b bs -> case deplacerBille labyrinthe deplacement bs b of
                                                    Just bi -> bi:bs
                                                    Nothing -> bs)
                                     []
                                     billes
        Nothing -> billes

-- Trie la liste des billes en mettant à droite les billes qui vont se déplacer devant
trierBilles :: Char -> [Coord] -> [Coord]
trierBilles 'N' billes = sortBy (\(y1,_) (y2,_) -> compare y2 y1) billes
trierBilles 'S' billes = sortBy (\(y1,_) (y2,_) -> compare y1 y2) billes
trierBilles 'E' billes = sortBy (\(_,x1) (_,x2) -> compare x2 x1) billes
trierBilles 'O' billes = sortBy (\(_,x1) (_,x2) -> compare x1 x2) billes

-- Déplace une bille dans une direction
deplacerBille :: Plan -> Coord -> [Coord] -> Coord -> Maybe Coord
deplacerBille labyrinthe deplacement billes position = do
    let suivant = (fst position + fst deplacement, snd position + snd deplacement)
    if labyrinthe!suivant == trou
        then Nothing
    else if labyrinthe!suivant == vide && not (suivant `elem` billes)
        then deplacerBille labyrinthe deplacement billes suivant
        else Just position
                                  
